tezos/timing/src/lib.rs

```diff
@@ -278,8 +282,61 @@ pub struct Action {
     pub tezedge_time: Option<f64>,
 }
 
-pub static TIMING_CHANNEL: Lazy<SyncSender<TimingMessage>> = Lazy::new(|| {
-    let (sender, receiver) = sync_channel(10_000);
+/// Buffered channel for sending timings that delays the sending until
+/// enough messages have been obtained or a commit message is received.
+/// The purpose is to send less messages through the channel to decrease
+/// the overhead.
+pub struct BufferedTimingChannel {
+    buffer: Mutex<Cell<Vec<TimingMessage>>>,
+    sender: SyncSender<Vec<TimingMessage>>,
+}
+
+impl BufferedTimingChannel {
+    const DELAYED_MESSAGES_LIMIT: usize = 100;
+
+    fn new(sender: SyncSender<Vec<TimingMessage>>) -> Self {
+        Self {
+            buffer: Mutex::new(Cell::new(Vec::with_capacity(Self::DELAYED_MESSAGES_LIMIT))),
+            sender,
+        }
+    }
+
+    /// True if the message must be sent immediately, false if it can be buffered.
+    fn is_immediate_message(&self, msg: &TimingMessage) -> bool {
+        match msg {
+            TimingMessage::Commit { .. } => true,
+            TimingMessage::InitTiming { .. } => false,
+            _ => false,
+        }
+    }
+
+    /// Sends messages, delayed and combined into a single bigger message.
+    ///
+    /// Reaching the delayed messages limit or receiving a commit message will trigger the send
+    /// to the underlying channel, otherwise the messages will be kept in the buffer.
+    pub fn send(&self, msg: TimingMessage) -> Result<(), SendError<Vec<TimingMessage>>> {
+        let must_not_delay = self.is_immediate_message(&msg);
+        let limit = Self::DELAYED_MESSAGES_LIMIT - 1;
+        let mut buffer = self.buffer.lock().unwrap();
+
+        buffer.get_mut().push(msg);
+
+        if must_not_delay || buffer.get_mut().len() == limit {
+            let swap_buffer = Cell::new(Vec::with_capacity(Self::DELAYED_MESSAGES_LIMIT));
+
+            buffer.swap(&swap_buffer);
+
+            let pack = swap_buffer.into_inner();
+
+            self.sender.send(pack)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub static TIMING_CHANNEL: Lazy<BufferedTimingChannel> = Lazy::new(|| {
+    let (sender, receiver) = sync_channel(1000);
```

Comment:
@tizoc 
is this enought? e.g. for big block with protocol switching?

Reply:
We actually have a bigger buffer now if you combine everything. The original buffer was 10k, now it is 10 time less but we are combining 100 messages into one, so thats 10x capacity (100k messages in total). Reverting it to 10k will not hurt tho.